### PR TITLE
Add Restore DNS options for our email services

### PR DESCRIPTION
**Pull request by daledupreez**
#### Changes proposed in this Pull Request

* This PR explores adding a general-purpose way to restore the correct DNS records for the email services we offer: Email Forwarding, Google Workspace/G Suite, and Professional Email
* As per discussion prompted by an early exploration, we're handling this via the restore/action menu on the DNS configuration page
* Also, as per [this comment](https://github.com/Automattic/wp-calypso/pull/60709#issuecomment-1028016982) and [this comment](https://github.com/Automattic/wp-calypso/pull/60709#issuecomment-1028007278), we're only showing the button if the user has that service and we definitively know that the user doesn't have the expected DNS records via the new flags added in D74409-code (we're also leveraging the new mail records-only template that was added in that change).
   - For now, the Titan menu option is disabled for subdomains, as we have some other bugs to address with how that template is handled for subdomains
* I also made a change to disable all of the restore buttons if the domain object is not currently known (which is most commonly the case while loading the page) -- this should help prevent unintentional DNS changes that aren't actually necessary.
* I am not sure how to create a good test case for Email Forwarding, as the API call to remove the email forwarding MX record also removes all email forwards

#### Testing instructions

* Run this branch locally or via [calypso.live](https://github.com/Automattic/wp-calypso/pull/60709#issuecomment-1026988290)
* For domains (or domain mappings) you own with various email services (i.e. Google Workspace/G Suite, Professional Email, or Email Forwarding), navigate to Upgrades -> Domains -> <your-domain.com>
* Expand the "DNS records" section, and click on the "Manage" button
* Click on the three-dot menu in the top right corner
* Verify that the Email Forwarding menu item is not visible
* Verify that the Professional Email menu item is not visible
* Verify that the Google Workspace menu item is not visible

Now, for domains that each have a subscription with Titan or Google, follow these steps:
* Navigate to Upgrades -> Domains -> <your-domain.com>
* Expand the "DNS records" section, and click on the "Manage" button
* Click on the three-dot menu in the top right corner
* Verify that all three email restore options are not visible
* Update the DNS to remove an `MX` record or remove a `TXT` record that starts with `v=spf1` and includes the email provider's service name
* Refresh the page to trigger a new API call to the domains end point
  - During one of your tests verify that none of the options in the top right menu, including the "Restore default A records" button, are visible while we're fetching data from the server
* Click on the three-dot menu in the top right corner
* Verify that the restore button for the current service is now visible, and click on it
* Verify that you see the confirmation dialog
* Verify that the record(s) you deleted are added back if you click on "Restore" and they are not if you click on "Cancel"
* Verify that clicking on either of the buttons results in the dialog being removed

For extra credit, verify that we disable the Professional Email option when dealing with a subdomain, even if it has a Professional Email subscription and invalid DNS records.

##### Screenshots

image
Screenshot 2022-02-02 at 11 47 08

### CODE REVIEW

**Comment by stephanethomas** [comment]
Quick comment but could we order providers alphabetically? That would make it easier to find what you're looking for.

**Comment by daledupreez** [comment]
> Quick comment but could we order providers alphabetically? That would make it easier to find what you're looking for.

Are we guaranteed to have the same sort order across languages? We may need to do this via a post-translation sort.

**Comment by delputnam** [comment]
Could we also please disable the restore records options if the records already exist? (Same as we do for the A records.)

**Comment by stephanethomas** [comment]
> Are we guaranteed to have the same sort order across languages? We may need to do this via a post-translation sort.

That is a very good point, I've never thought of that.

**Comment by delputnam** [comment]
I'm also thinking that perhaps it's best to not show the email restore options at all unless the service is attached to the domain in the first place.  So, for example, if you don't have any email forwards set up, then don't even show the restore option for the email forwards.  But if you have a forward and the correct record is present, then show the option disabled.  And if you have a forward, but the record is not present, then show the option enabled.

I'm worried that users who see the options, but don't have the services will be confused.

**Comment by daledupreez** [comment]
> I'm also thinking that perhaps it's best to not show the email restore options at all unless the service is attached to the domain in the first place. So, for example, if you don't have any email forwards set up, then don't even show the restore option for the email forwards. But if you have a forward and the correct record is present, then show the option disabled. And if you have a forward, but the record is not present, then show the option enabled.
> 
> I'm worried that users who see the options, but don't have the services will be confused.

I think there are multiple complications on this front:
* When we enable a paid email service, we take two sets of actions that cause trouble for us trying to work out whether you should (still) have email forwarding records:
   1. We update the DNS to add the Google/Titan DNS records, and remove the previous email service records
   2. We queue up deletion of any existing email forwards
* I am not sure that we have a good (or at least light-weight) way of determining whether the records for a given template have been applied, and if we did, I am not sure where that data should be exposed -- it might need to be in a new API

**Comment by delputnam** [comment]
> I think there are multiple complications on this front:
> 
> * When we enable a paid email service, we take two sets of actions that cause trouble for us trying to work out whether you should (still) have email forwarding records:
>   
>   1. We update the DNS to add the Google/Titan DNS records, and remove the previous email service records
>   2. We queue up deletion of any existing email forwards
> * I am not sure that we have a good (or at least light-weight) way of determining whether the records for a given template have been applied, and if we did, I am not sure where that data should be exposed -- it might need to be in a new API

We should be able to add the checks for the email records in to the `/site/xxx/domains` endpoint. It should be a very fast check. If you want, I can draft a patch for this later in the week.

I think that even though there will be a short window of time where it might look like there are both forwards and some other service active, this should be ok. Since forwards and any other service are mutually exclusive, we can just not show the option to restore the forwarding record if one of the other services is active.

**Comment by daledupreez** [comment]
I've started exploring adding those fields to the `/site/:blog_id/domains` API in D74409-code, which also adds a Gmail template that doesn't require a token. I'll update this logic to match with the data once we confirm the shape of the data.

**Comment by delputnam** [comment]
I still think we should not even show an option to a user who doesn't also have a subscription to the corresponding service. I think it is confusing to show them for these cases.

Also, I hadn't thought of this before, but as you pointed out, when the forwarding record is remove, we remove the forwards, so it should be very rare (and would indicate a bug) to have the case where a user has an email forward, but no corresponding MX record.  This means that there would never be a case where it makes sense to even show the option.  The correct fix here it for the user to go to the email management page and add forwards again.

So, I think this option really only makes sense for Titan and Google...and only for users who have one of those subscriptions.

**Comment by daledupreez** [comment]
> I still think we should not even show an option to a user who doesn't also have a subscription to the corresponding service. I think it is confusing to show them for these cases.

I can modify the logic to avoid showing the options altogether for these two cases. From my side, I almost always err towards showing a disabled option for something so that a user knows where to find it when they might need it, rather than it being a surprise in the handful of situations where it suddenly does become useful. But I am happy to defer to your preference on this one.

> Also, I hadn't thought of this before, but as you pointed out, when the forwarding record is remove, we remove the forwards, so it should be very rare (and would indicate a bug) to have the case where a user has an email forward, but no corresponding MX record. This means that there would never be a case where it makes sense to even show the option.

I'll remove the handling for the email forwarding option in this case. As you mention, this situation would likely be an extreme edge case.

**Comment by daledupreez** [comment]
@delputnam, I've worked in the changes to hide the menu items when they are not applicable, but that really does feel like a bad UX to me. As I've sat with it for just over 10 minutes, I have a stronger resistance than I thought! My biggest concern is that we're indirectly training users not to look at this menu for the feature, because if the user does click on this menu during normal interactions, we're not giving them any indication that email restore options might appear or be available in future. I worry that while we may have the button present, we may still end up with users contacting support.

**Comment by AllTerrainDeveloper** [comment]
Wow, regarding showing disabled options or not showing them at all I have mixed feelings... But I am leaning to remove them, and this is my opinion and why:

I would say that yes, it might be confusing for users to see something they don't even know what is it.
 - _Professional Email? How can I enable that button? What I did wrong? Why is disabled?_

A good thing is that we are showing a brand, giving them presence. But the bad thing is that we are showing both together. I can understand that we present both options in the Emails comparison page, that's completely needed and fair. But if we present both options here, to fix/repair things we are reminding the user that there are more options for emails (and that they might fail) and maybe it wouldn't like either to Titan or Google to be next to each other.

In the other side, more experienced and bold users, would find this disabled option useful and remember where is it when they have some kind of problem with DNS.

BUT! as a personal opinion it seems to me that is quite a bit hidden, and we should show this as a button/link near/above/below the domain that is having problems, not having the correct DNS. Some warning, some collapsable information instead of a popup shown through a ellipsis menu, which they also are sometimes annoying in mobile devices.

File: client/lib/domains/types.ts [hunk 1]
```diff
@@ -104,7 +106,7 @@ export type ResponseDomain = {
 	registryExpiryDate: string;
 	sslStatus: string | null;
 	subscriptionId: string | null;
-	subdomainPart: string;
+	subdomainPart?: string;
```
Comment (AllTerrainDeveloper) [inline client/lib/domains/types.ts:108]: Super NIT: Should not go `subdomainPart` before `subscriptionId`? (alphabetically speaking)

**Comment by AllTerrainDeveloper** [comment]
@daledupreez In the code I see that we disable the button, but the testing plan is old, and asks for checking the menu item visibility 🙂

**Review by AllTerrainDeveloper — APPROVED** [review]
Looks good to me! Although during testing it felt strange to me that I had to refresh page to see the enabled option, but I recon that user would not delete a record and immediately restore the DNS config 😆 

### EOT

**Professional Email Test**

Show option when Professional email have a wrong configuration
image

The dialog did appear and I clicked in restore.

The DNS were fixed.

**Google Workspace Test**

Show option when Google Workspace has wrong DNS config
image

The dialog did appear and I clicked in restore.

The DNS were fixed.

**Comment by delputnam** [comment]
The reasons that @AllTerrainDeveloper mentioned are exactly the reason why I think we shouldn't show these specific options at all when they don't apply.  I think that users who see them and don't have a subscription and maybe have never had a subscription to these products will think that there is something wrong and be confused or frustrated.

I had also thought about product visibility, but ultimately concluded that action menu options shouldn't be treated as an advertisement for a product.

**Comment by daledupreez** [comment]
Thanks for the reviews and the discussion, @delputnam and @AllTerrainDeveloper! Much appreciated!

And I think it may be better to keep these links for when we need them, especially if we can get them documented.

File: client/lib/domains/types.ts (hunk 1 above)
Comment (daledupreez) [reply client/lib/domains/types.ts:108]: Good catch! I just fixed this.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7140792

Thank you @daledupreez for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7140792).

**Pull request merged**